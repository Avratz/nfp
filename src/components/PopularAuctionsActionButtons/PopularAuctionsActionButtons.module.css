.auctionButtons {
  display: flex;
  flex-direction: column;
  width: 100%;
  gap: 0.5rem;
}

.modal {
  animation: var(--fadeIn);
  background-color: rgba(20, 20, 22, 0.95);
  box-shadow: 0 0.25rem 0.5rem var(--neutral-1);
  backdrop-filter: blur(7px);
}

.modal,
.success {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 2rem;
  padding: 6rem 2rem;
  border-radius: 1rem;
  background-color: var(--neutral-1);
  width: 90%;
  max-width: 500px;
  position: relative;
}

.success {
  position: absolute;
  top: 0;
  left: 0;
  background-color: var(--green);
  width: 100%;
  height: 100%;
  z-index: 99;
}

.modalText {
  font-family: var(--font-poppins);
  font-weight: var(--font-medium);
  font-size: var(--text-xs);
  line-height: var(--line-xs);
  color: var(--neutral-4);
  margin-bottom: 2rem;
  animation: var(--fadeInLong);
}

.modalButtons {
  display: flex;
  justify-content: center;
  width: 100%;
  gap: 1rem;
}

.closeIcon {
  position: absolute;
  top: 1rem;
  right: 1rem;
  cursor: pointer;
}

.finishButton {
  border-color: var(--neutral-8);
}
.checkmarkIconWrapper {
  width: 100px;
  height: 100px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.checkmarkCircle {
  stroke-dasharray: 166;
  stroke-dashoffset: 166;
  stroke-width: 2;
  stroke-miterlimit: 10;
  stroke: var(--neutral-8);
  fill: none;
  animation: stroke 0.6s cubic-bezier(0.65, 0, 0.45, 1) forwards;
}

.checkmark {
  width: 56px;
  height: 56px;
  border-radius: 50%;
  display: block;
  stroke-width: 3;
  stroke: var(--green);
  stroke-miterlimit: 10;
  margin: 10% auto;
  box-shadow: inset 0px 0px 0px var(--neutral-8);
  animation: fill 0.4s ease-in-out 0.4s forwards,
    scale 0.3s ease-in-out 0.9s both;
}

.checkmarkCheck {
  transform-origin: 50% 50%;
  stroke-dasharray: 48;
  stroke-dashoffset: 48;
  animation: stroke 0.3s cubic-bezier(0.65, 0, 0.45, 1) 0.8s forwards;
}

@keyframes stroke {
  100% {
    stroke-dashoffset: 0;
  }
}
@keyframes scale {
  0%,
  100% {
    transform: none;
  }
  50% {
    transform: scale3d(1.1, 1.1, 1);
  }
}
@keyframes fill {
  100% {
    box-shadow: inset 0px 0px 0px 30px var(--neutral-8);
  }
}
