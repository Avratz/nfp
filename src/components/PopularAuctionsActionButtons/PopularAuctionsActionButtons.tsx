import { FormEvent, Fragment, useRef, useState } from 'react'
import { motion } from 'framer-motion'
import dynamic from 'next/dynamic'

import { useOverlay } from '@/hooks/useOverlay'

import { Button } from '@/ui/Button/Button'

const Modal = dynamic(
  () => import('@/ui/Modal/Modal').then((mod) => mod.Modal),
  { ssr: false },
)

import { Heading } from '@/ui/Heading/Heading'
import { Icon } from '@/ui/Icon/Icon'
import { Input } from '@/ui/Input/Input'

import styles from './PopularAuctionsActionButtons.module.css'
import { newAudio } from '@/lib/audio'

const success = {
  open: {
    clipPath: `circle(100% at 50% 50%)`,
    transition: {
      type: 'spring',
      stiffness: 20,
    },
  },
  closed: {
    clipPath: 'circle(0px at 50% 50%)',
    transition: {
      type: 'spring',
    },
  },
}

export function PopularAuctionsPlaceNewBid({
  highestBid,
}: {
  highestBid: string
}) {
  const { isOpen, setIsOpen } = useOverlay()
  const [isSuccess, setIsSuccess] = useState(false)

  function confirmBid(event: FormEvent<HTMLFormElement>) {
    event.preventDefault()
    const formData = new FormData(event.currentTarget)
    const newBid = formData.get('bid') as string
    if (newBid === '') return
    setIsSuccess(true)
    newAudio('/audio/success.mp3').play()
  }

  function closeModal() {
    setIsOpen(false)
    setIsSuccess(false)
  }

  return (
    <div className={styles.auctionButtons}>
      {isOpen ? (
        <Modal closeModal={closeModal}>
          <motion.form
            className={styles.modal}
            onClick={(e) => e.stopPropagation()}
            onSubmit={confirmBid}
          >
            {isSuccess ? (
              <motion.div
                className={styles.success}
                initial={'closed'}
                animate={isSuccess ? 'open' : 'closed'}
                variants={success}
              >
                <div className={styles.checkmarkIconWrapper}>
                  <svg
                    className={styles.checkmark}
                    xmlns="http://www.w3.org/2000/svg"
                    viewBox="0 0 52 52"
                  >
                    <circle
                      className={styles.checkmarkCircle}
                      cx="26"
                      cy="26"
                      r="25"
                      fill="none"
                    />
                    <path
                      className={styles.checkmarkCheck}
                      fill="none"
                      d="M14.1 27.2l7.1 7.2 16.7-16.8"
                    />
                  </svg>
                </div>

                <Heading as="h3" size="2xl">
                  Bid placed successfully!
                </Heading>

                <Button
                  intent="primary"
                  className={styles.finishButton}
                  onClick={closeModal}
                  type="button"
                >
                  Close
                </Button>
              </motion.div>
            ) : (
              <Fragment>
                <Button
                  intent="icon"
                  onClick={closeModal}
                  className={styles.closeIcon}
                  title="Close modal"
                  type="button"
                >
                  <Icon name="closeCircle" size="medium" />
                </Button>
                <Heading as="h3" size="2xl">
                  Place a new bid
                </Heading>
                <p className={styles.modalText}>
                  The current bid is {highestBid}
                </p>
                <Input type="number" name="bid" required />
                <div className={styles.modalButtons}>
                  <Button intent="accent">Confirm</Button>
                  <Button onClick={closeModal} type="button">
                    Cancel
                  </Button>
                </div>
              </Fragment>
            )}
          </motion.form>
        </Modal>
      ) : null}
      <Button intent="accent" onClick={() => setIsOpen(true)}>
        Place a bid
      </Button>
      <Button intent="primary">View item</Button>
    </div>
  )
}
