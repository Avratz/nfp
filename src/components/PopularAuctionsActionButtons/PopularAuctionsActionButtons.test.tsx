import { render } from '@testing-library/react'
import { PopularAuctionsActionButtons } from './PopularAuctionsActionButtons'

jest.mock('@/ui/Icon/Icon', () => {
  function Icon() {
    return <div></div>
  }

  Icon.displayName = 'Icon'
  return {
    __esModule: true,
    Icon,
  }
})
describe('PopularAuctionsActionButtons', () => {
  it('renders PopularAuctionsActionButtons', () => {
    let { container } = render(<PopularAuctionsActionButtons highestBid="50" />)
    expect(container).toMatchSnapshot()
  })
})
