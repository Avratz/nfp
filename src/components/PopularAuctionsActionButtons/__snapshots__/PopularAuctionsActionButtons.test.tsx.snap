// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopularAuctionsActionButtons renders PopularAuctionsActionButtons 1`] = `
<div>
  <div
    class="auctionButtons"
  >
    <button
      class="button accent base"
    >
      Place a bid
    </button>
    <button
      class="button primary base"
    >
      View item
    </button>
  </div>
</div>
`;
